gstd/src/common/errors.rs

```diff
@@ -33,50 +34,77 @@ pub type Result<T, E = Error> = core::result::Result<T, E>;
-    ReplyCode(ReplyCode),
-    /// This error occurs when providing zero duration to waiting functions
-    /// (e.g. see `exactly` and `up_to` functions in
-    /// [CodecMessageFuture](crate::msg::CodecMessageFuture)).
-    EmptyWaitDuration,
-    /// This error occurs when providing zero gas amount to system gas reserving
-    /// function (see
-    /// [Config::set_system_reserve](crate::Config::set_system_reserve)).
-    ZeroSystemReservationAmount,
-    /// This error occurs when providing zero duration to mutex lock function
-    ZeroMxLockDuration,
+
+    /// Gstd API usage error.
+    ///
+    /// Note: this error returns from `gstd` APIs in case of invalid arguments.
+    Gstd(UsageError),
+
+    /* Business logic errors */
+    /// Received error reply while awaited response from another actor.
+    ///
+    /// NOTE: this error could only be returned from async messaging.
+    ErrorReply(ErrorReplyPayload, ErrorReplyReason),
+
+    /// Received reply that couldn't be identified as successful or not
+    /// due to unsupported reply code.
+    ///
+    /// NOTE: this error could only be returned from async messaging.
+    UnsupportedReply(Vec<u8>),
+
+    /// Timeout reached while expecting for reply.
+    ///
+    /// NOTE: this error could only be returned from async messaging.
+    Timeout(u32, u32),
 }
 
 impl Error {
     /// Check whether an error is [`Error::Timeout`].
     pub fn timed_out(&self) -> bool {
         matches!(self, Error::Timeout(..))
     }
+
+    /// Check whether an error is [`Error::ErrorReply`] and return its str
+    /// representation.
+    pub fn error_reply_str(&self) -> Option<&str> {
+        if let Self::ErrorReply(payload, _) = self {
+            payload.as_str()
+        } else {
+            None
+        }
+    }
 }
 
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
             Error::Core(e) => fmt::Display::fmt(e, f),
-            Error::Timeout(expected, now) => {
-                write!(f, "Wait lock timeout at {expected}, now is {now}")
-            }
             Error::Convert(e) => write!(f, "Conversion error: {e:?}"),
-            Error::Decode(e) => write!(f, "Decoding codec bytes error: {e}"),
-            Error::ReplyCode(e) => write!(f, "Reply came with non success reply code {e:?}"),
-            Error::EmptyWaitDuration => write!(f, "Wait duration can not be zero."),
-            Error::ZeroSystemReservationAmount => {
-                write!(f, "System reservation amount can not be zero in config.")
+            Error::Decode(e) => write!(f, "Scale codec decoding error: {e}"),
+            Error::Gstd(e) => write!(f, "`Gstd` API error: {e:?}"),
+            Error::ErrorReply(err, reason) => write!(f, "Received reply '{err}' due to {reason:?}"),
+            Error::UnsupportedReply(payload) => {
+                write!(f, "Received unsupported reply 0x'{}'", hex::encode(payload))
```

Comment:
to match with `impl Display for ErrorReplyPayload`
```suggestion
                write!(f, "Received unsupported reply '0x{}'", hex::encode(payload))
```